### ISSUE TITLE
[CUDA][Performance] YOLOX-nano training loop is ~36.5x slower than equivalent PyTorch minimal repro 

### ISSUE THREAD

**Issue by JacobCote** [tracel-ai/burn#4598]
## Summary

I observe a very large performance gap between Burn CUDA and PyTorch CUDA on an equivalent **minimal synthetic training loop**:

- Burn median step: **1239.90 ms**
- PyTorch median step: **33.97 ms**
- Burn is about **36.5x slower**

This reproduces with:
- no dataset / dataloader
- fixed zero input tensor
- simple loss
- train loop only (`forward -> loss -> backward -> optimizer.step`)

## Environment
- GPU: NVIDIA Quadro RTX 6000
- Driver: 580.126.09
- NVIDIA-SMI CUDA Version: 13.0
- nvcc --version: 12.6 (V12.6.85)
- rustc --version: rustc 1.93.0 (254b59607 2026-01-19)
- Burn crate: 0.20.0
- yolox-burn: https://github.com/tracel-ai/models (Cargo resolved rev 4bca43f3)
- yolox-pytorch : https://github.com/Megvii-BaseDetection/YOLOX
---

### 1) Burn CUDA minimal repro

```rust
#![recursion_limit = "256"]

use burn::backend::cuda::CudaDevice;
use burn::backend::{Autodiff, Cuda};
use burn::optim::{AdamConfig, GradientsParams, Optimizer};
use burn::tensor::backend::Backend;
use burn::tensor::Tensor;
use yolox_burn::model::yolox::Yolox;

type B = Autodiff<Cuda>;

const BATCH_SIZE: usize = 4;
const IMAGE_SIZE: usize = 416;
const STEPS: usize = 20;
const LR: f64 = 1e-4;
const NUM_CLASSES: usize = 1;

fn main() {
    let device = CudaDevice::default();
    let mut model = Yolox::<B>::yolox_nano(NUM_CLASSES, &device);
    let mut optimizer = AdamConfig::new().init();

    let input = Tensor::<B, 4>::zeros([BATCH_SIZE, 3, IMAGE_SIZE, IMAGE_SIZE], &device);

    let mut last_loss = 0.0f32;
    for _ in 0..STEPS {
        let output = model.forward(input.clone());
        let loss = (output.clone() * output).mean();
        let grads = loss.clone().backward();
        let grads = GradientsParams::from_grads(grads, &model);
        model = optimizer.step(LR, model, grads);
        last_loss = loss.detach().into_scalar();
    }

    B::sync(&device).unwrap();
    println!("last_loss={}", last_loss);
}
```
### 2) Pytorch CUDA  minimal repro

```python 

BATCH_SIZE = 4
IMAGE_SIZE = 416
STEPS = 20
LR = 1e-4
NUM_CLASSES = 1
DEVICE = torch.device("cuda:0")

def build_yolox_nano(num_classes):
    backbone = YOLOPAFPN(
        depth=0.33,
        width=0.25,
        in_features=("dark3", "dark4", "dark5"),
        in_channels=[256, 512, 1024],
        depthwise=True,
        act="silu",
    )
    head = YOLOXHead(
        num_classes=num_classes,
        width=0.25,
        strides=[8, 16, 32],
        in_channels=[256, 512, 1024],
        depthwise=True,
        act="silu",
    )
    model = YOLOX(backbone=backbone, head=head)
    model.head.initialize_biases(1e-2)
    return model

def main():
    if not torch.cuda.is_available():
        raise RuntimeError("CUDA not available")

    model = build_yolox_nano(NUM_CLASSES).to(DEVICE)
    model.eval()  # model returns predictions directly in eval mode
    opt = torch.optim.Adam(model.parameters(), lr=LR)
    x = torch.zeros(BATCH_SIZE, 3, IMAGE_SIZE, IMAGE_SIZE, device=DEVICE)

    loss = None
    for _ in range(STEPS):
        opt.zero_grad(set_to_none=True)
        out = model(x)
        loss = (out * out).mean()
        loss.backward()
        opt.step()

    torch.cuda.synchronize(DEVICE)
    print(f"last_loss={float(loss.item())}")

if __name__ == "__main__":
    main()

```

**Comment by laggui** [comment]
Thanks for the report.

The performance might not always be on par for vision models, but the gap is probably not as big.

A couple of things to note:
- the first steps will be heavily influenced by the autotuning, which are typically discarded as "warm-up" when benchmarking
- `sync` should not only be called once outside the loop, but for each step
- `model.eval()` also changes the behavior (e.g., disables batchnorm updates)

The number of steps should probably be a bit higher to get a more meaningful comparison, excluding warm-up.

**Comment by antimora** [comment]
Also double check if it's built with `--release` it's a common mistake too

**Comment by JacobCote** [comment]
@laggui Thanks for your response. I was only showing a minimalist repro for the example. Here is the full training + logging loop I used.  So 20 warm up steps + sync at each step. I also re-ran the PyTorch benchmark with mode.train() and fake targets. 

here are the PyTorch results with .train():

#### Summary:
  - mean step:   42.01 ms
  - median step: 42.00 ms
  - p95 step:    42.34 ms

 #### median breakdown: fwd 18.86 | loss 0.39 | bwd 17.87 | opt 4.83


The previous rust benchmark was run with: 

~~~bash
cargo run --release --no-default-features --features cuda --bin yolox_min_repro
~~~

Here are the burn results: 
#### Summary:
 - mean step:   1268.23 ms
 - median step: 1262.61 ms
 - p95 step:    1289.43 ms
 #### median breakdown: fwd 207.23 | loss 0.12 | bwd 956.38 | opt 99.62

~~~rust

#![recursion_limit = "256"]

#[cfg(feature = "cuda")]
fn main() {
    use burn::backend::cuda::CudaDevice;
    use burn::backend::{Autodiff, Cuda};
    use burn::module::Module;
    use burn::optim::{AdamConfig, GradientsParams, Optimizer};
    use burn::tensor::Tensor;
    use burn::tensor::backend::Backend;
    use std::time::Instant;
    use yolox_burn::model::yolox::Yolox;

    type MyBackend = Autodiff<Cuda>;

    const BATCH_SIZE: usize = 4;
    const IMAGE_SIZE: usize = 416;
    const WARMUP_STEPS: usize = 20;
    const MEASURE_STEPS: usize = 20;
    const LEARNING_RATE: f64 = 1e-4;
    const NUM_CLASSES: usize = 1;

    #[derive(Clone, Copy)]
    struct StepTiming {
        total_ms: f64,
        forward_ms: f64,
        loss_ms: f64,
        backward_ms: f64,
        opt_ms: f64,
    }

    fn sync(device: &CudaDevice) {
        MyBackend::sync(device).expect("cuda sync should succeed");
    }

    fn mean(values: &[f64]) -> f64 {
        values.iter().sum::<f64>() / values.len() as f64
    }

    fn median(values: &[f64]) -> f64 {
        let mut values = values.to_vec();
        values.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
        let mid = values.len() / 2;
        if values.len() % 2 == 0 {
            (values[mid - 1] + values[mid]) * 0.5
        } else {
            values[mid]
        }
    }

    fn p95(values: &[f64]) -> f64 {
        let mut values = values.to_vec();
        values.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
        let idx = ((values.len() as f64) * 0.95).ceil() as usize;
        values[idx.saturating_sub(1).min(values.len() - 1)]
    }

    let device = CudaDevice::default();
    let mut model = Yolox::<MyBackend>::yolox_nano(NUM_CLASSES, &device);
    let mut optimizer = AdamConfig::new().init();
    let input = Tensor::<MyBackend, 4>::zeros([BATCH_SIZE, 3, IMAGE_SIZE, IMAGE_SIZE], &device);

    println!("Using device: {:?}", device);
    println!("Trainable parameters: {}", model.num_params());
    println!(
        "Config: batch_size={} image_size={} warmup_steps={} measure_steps={}",
        BATCH_SIZE, IMAGE_SIZE, WARMUP_STEPS, MEASURE_STEPS
    );

    let mut timings = Vec::with_capacity(MEASURE_STEPS);

    for step in 0..(WARMUP_STEPS + MEASURE_STEPS) {
        sync(&device);
        let forward_start = Instant::now();
        let output = model.forward(input.clone());
        sync(&device);
        let forward_ms = forward_start.elapsed().as_secs_f64() * 1000.0;

        let loss_start = Instant::now();
        let loss = (output.clone() * output).mean();
        sync(&device);
        let loss_ms = loss_start.elapsed().as_secs_f64() * 1000.0;

        let backward_start = Instant::now();
        let grads = loss.clone().backward();
        let grads = GradientsParams::from_grads(grads, &model);
        sync(&device);
        let backward_ms = backward_start.elapsed().as_secs_f64() * 1000.0;

        let opt_start = Instant::now();
        model = optimizer.step(LEARNING_RATE, model, grads);
        sync(&device);
        let opt_ms = opt_start.elapsed().as_secs_f64() * 1000.0;

        let loss_value = loss.detach().into_scalar();
        let total_ms = forward_ms + loss_ms + backward_ms + opt_ms;

        if step >= WARMUP_STEPS {
            timings.push(StepTiming {
                total_ms,
                forward_ms,
                loss_ms,
                backward_ms,
                opt_ms,
            });
            println!(
                "step {} - loss {:.6} | total {:.1} ms | fwd {:.1} | loss {:.1} | bwd {:.1} | opt {:.1}",
                step + 1 - WARMUP_STEPS,
                loss_value,
                total_ms,
                forward_ms,
                loss_ms,
                backward_ms,
                opt_ms
            );
        }
    }

    let total: Vec<f64> = timings.iter().map(|t| t.total_ms).collect();
    let forward: Vec<f64> = timings.iter().map(|t| t.forward_ms).collect();
    let loss: Vec<f64> = timings.iter().map(|t| t.loss_ms).collect();
    let backward: Vec<f64> = timings.iter().map(|t| t.backward_ms).collect();
    let opt: Vec<f64> = timings.iter().map(|t| t.opt_ms).collect();

    println!();
    println!("Summary:");
    println!("  mean step:   {:.2} ms", mean(&total));
    println!("  median step: {:.2} ms", median(&total));
    println!("  p95 step:    {:.2} ms", p95(&total));
    println!(
        "  median breakdown: fwd {:.2} | loss {:.2} | bwd {:.2} | opt {:.2}",
        median(&forward),
        median(&loss),
        median(&backward),
        median(&opt)
    );
}

~~~

**Comment by laggui** [comment]
> Thanks for your response. I was only showing a minimalist repro for the example. Here is the full training + logging loop I used. So 20 warm up steps + sync at each step. I also re-ran the PyTorch benchmark with mode.train() and fake targets.

Ahh thanks for providing more info! The benchmark setup wasn't entirely clear from the initial report.

The backward is orders of magnitudes slower than the forward, which is suspicious.

This might be a combination of factors and unoptimized paths. At the implementation level, I see interpolate as potential issue, and inefficient decode implementation. Convolution backward implementation should have improved with recent changes. Would have to profile the model to identify the biggest culprits.
